packages/app/src/runner/event-manager.ts

```diff
@@ -361,19 +360,31 @@ export class EventManager {
   }
 
   setup (config) {
-    Cypress = this.Cypress = this.$CypressDriver.create(config)
+    this.ws.emit('watch:test:file', config.spec)
 
-    // expose Cypress globally
-    // @ts-ignore
-    window.Cypress = Cypress
+    return new Promise((resolve, reject) => {
+      if (!config.isTextTerminal && !config.experimentalInteractiveRunEvents) {
+        return resolve(null)
+      }
 
-    this._addListeners()
+      this.ws.emit('plugins:before:spec', config.spec, (res?: { error: Error }) => {
+        // FIXME: handle surfacing the error to the browser instead of hanging with
+        // 'Your tests are loading...' message. Fix in https://github.com/cypress-io/cypress/issues/23627
+        if (res && res.error) {
+          reject(res.error)
+        }
 
-    this.ws.emit('watch:test:file', config.spec)
+        resolve(null)
+      })
+    }).then(() => {
```

Comment:
You can (probably) just use `async/await` here (if you want, that is - I don't think this part is impacted by the whole "async doesn't work properly" since this is outside of the driver, but either way is fine by me.

Reply:
The `async won't work properly` concept is tied to writing Cypress tests, not the code so await should work fine.